pgxn/neon/pagestore_smgr.c

```diff
@@ -236,14 +255,112 @@ static void prefetch_read(PrefetchRequest *slot);
 static void prefetch_do_request(PrefetchRequest *slot, bool *force_latest, XLogRecPtr *force_lsn);
 static void prefetch_wait_for(uint64 ring_index);
 static void prefetch_cleanup(void);
-static inline void prefetch_set_unused(uint64 ring_index, bool hash_cleanup);
+static inline void prefetch_set_unused(uint64 ring_index);
 
 static XLogRecPtr neon_get_request_lsn(bool *latest, RelFileNode rnode,
 									   ForkNumber forknum, BlockNumber blkno);
 
+void
```

Comment:
> No, because the storage manager doesn't (and can't/shouldn't) know when the query is finished.

It is not necessary to know when query is finished.
You can just delay applying config changes until prefetch buffer is empty. And it can be checked in neon_prefetch  function.

Reply:
Sorry, but I am not convinced with your arguments.
Any prefetch requests should be completed until end of query execution.
Yes query execution itself can take a long time: minute, hours or may be even days...
The question is whether we need to be able to change prefetch parameter during query execution?
It can be done only using ALTER SYSTEM and `pg_reload_conf()`. I wonder if Neon user are ever have such privilege.
If not, then you are trying to solve unexisted problem.

But even if it is possible, it is quite questionable whether user really want to change prefetch parameters during query execution. And i any case: it is very error prone and dangerous. So I prefer not to address all this possible issues and not to write complex code with doing some doubtful job. But it is certainly your choice.

Reply:
If we make it USERSET, we certainly have to be prepared for it being changed in the middle of a statement. It doesn't need to be very fast or clever though, I wouldn't expect anyone to actually change it on the fly, the point of making it USERSET is to make it easier to test different values.

So we don't necessarily need to keep the already-prefetched requests. We could just disconnect the connection to the pageserver, and free and reinitialize the queue from scratch.

The code here looks OK to me, too, though.